openfl/component/straggler_handling_functions/cutoff_time_based_straggler_handling.py

```diff
@@ -4,30 +4,130 @@
+        # the timer again until next round.
+        if self.straggler_cutoff_time == np.inf or self.__is_policy_applied_for_round:
+            return
+        self.round_start_time = time.time()
+        if hasattr(self, "timer"):
+            self.timer.cancel()
+            delattr(self, "timer")
+        self.timer = threading.Timer(
+            self.straggler_cutoff_time, callback,
+        )
+        self.timer.daemon = True
+        self.timer.start()
 
-    def straggler_cutoff_check(self, num_collaborators_done, all_collaborators=None):
-        cutoff = self.straggler_time_expired() and self.minimum_collaborators_reported(
-            num_collaborators_done)
-        return cutoff
+    def straggler_cutoff_check(
+        self, num_collaborators_done: int, num_all_collaborators: int,
+    ) -> bool:
+        """
+        If minimum_reporting collaborators have reported results within
+        straggler_cutoff_time then return True, otherwise False.
+
+        Args:
+            num_collaborators_done: int
+                Number of collaborators finished.
+            num_all_collaborators: int
+                Total number of collaborators.
+
+        Returns:
+            bool
+        """
+        if not self.__straggler_time_expired():
+            return False
+        # Check if time has expired and policy is not applied
+        elif self.__straggler_time_expired() and not self.__is_policy_applied_for_round:
+            # Stop timer from restarting for current round and
+            # if set to True wait for ALL collaborators instead of minimum_reporting
+            self.__is_policy_applied_for_round = True
+            # Check if minimum_reporting collaborators have reported results
+            if self.__minimum_collaborators_reported(num_collaborators_done):
+                self.logger.info(
+                    f"{num_collaborators_done} collaborators reported results within "
+                    "cutoff time. Applying cutoff policy and proceeding with end of round."
+                )
+                return True
+            self.logger.info(
+                "Disregarding straggler handling policy and waiting for ALL "
+                f"{num_all_collaborators} collaborator(s) to report results."
+            )
+            return False
+        # If straggler_cutoff_time is set to infinite,
+        # OR
+        # If minimum_reporting collaborators have not reported results within cutoff
+        # time,
+        # then disregard the policy and wait for ALL collaborators to report
+        # results.
+        elif (
+            self.straggler_cutoff_time == np.inf
+            or (self.__straggler_time_expired() and self.__is_policy_applied_for_round)
+        ):
+            return num_all_collaborators == num_collaborators_done
+        # Something has gone, unhandled scenario, raising error.
+        raise ValueError("Unhandled scenario")
```

Comment:
for debugging purposes it would be useful to also log the parameters of the `if` statements above:
```python
self.__straggler_time_expired()
self.__is_policy_applied_for_round
self.straggler_cutoff_time
```

Reply:
As you suggested, I have added 3 following statements in error message:

```
self.__straggler_time_expired()
self.__is_policy_applied_for_round
self.straggler_cutoff_time
```